ibis/expr/operations/core.py

```diff
@@ -23,7 +24,12 @@ def equals(self, other) -> bool:
             raise TypeError(
                 f"invalid equality comparison between Node and {type(other)}"
             )
-        return self.__cached_equals__(other)
+
+        lefts = Graph.from_bfs(self).toposort()
```

Comment:
This is pretty aggressive to me. Topological sort is not a cheap algorithm, and we shouldn't optimize ibis for unreasonable usage. Besides that  this is inconsistent with `self.op() == other.op()` which would trigger the original code path.

Reply:
> This is pretty aggressive to me. Topological sort is not a cheap algorithm, and we shouldn't optimize ibis for unreasonable usage

Agree, though it's hard to say what a ... reasonable definition of unreasonable is 😄 

> Besides that this is inconsistent with self.op() == other.op() which would trigger the original code path

How are verifying this? It doesn't seem to be true.

Reply:
`self.op() == other.op()` calls `Node.__eq__` which does the equality check recursively.